File: src/LinearRegression/LinearRegression.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <utility>
#include <cmath>
using namespace std;


// Function to perform linear regression analyisis
template <class T>
pair<float, float> linearRegression(vector<T>& xs, vector<T>& ys) {
    /* 
        This function takes in 2 vectors of ints or floats representing x and y respectively.
        And returns a pair containing the gradient (m) in the first and y-intercept (b) in the second
        
        The formula used is from Khan Academy - https://www.youtube.com/watch?v=GAmzwIkGFgE
     */
    
    // Size of xs and ys
    int N_x = xs.size();
    int N_y = ys.size();

    // Calculate average of x and y
    float avg_x = accumulate(xs.begin(), xs.end(), 0) / N_x; 
    float avg_y = accumulate(ys.begin(), ys.end(), 0) / N_y;

    // Average of xy's
    vector<float> xys;

    for (int i = 0; i < N_x; ++i) {
        xys.push_back( xs[i] * ys[i] );
    }

    float avg_xy = accumulate(xys.begin(), xys.end(), 0) / N_x;

    // Calculate the square of average x
    float squared_avg_x = pow(avg_x, 2);

    // Calculate the average of square of xs
    vector<float> xs_squared;

    for (int j = 0; j < N_x; ++j) {
        xs_squared.push_back( pow(xs[j], 2) );
    }

    float avg_xs_squared = accumulate(xs_squared.begin(), xs_squared.end(), 0) / N_x;

    // Calculate the gradient
    float m = ((avg_x * avg_y) - avg_xy) / (squared_avg_x - avg_xs_squared);
    float b = avg_y - (m * avg_x);

    // Return {gradient, y-intersect}
    return {m, b};
};

int main() {

    // This dataset is has a positive correlation with the y-intersect being close to the origin (0,0)
    vector<int> x = {1, 2, 4, 6,  8, 10, 15, 17};
    vector<int> y = {1, 3, 2, 6, 11, 10, 13, 20};
    
    pair<float, float> equation = linearRegression(x, y);

    cout << "m = " << equation.first << endl;
    cout << "b = " << equation.second << endl;

    return 0;
}

File: src/LinearRegression/LinearRegression.py


# Function to perform linear regression analysis
def linearRegression(xs, ys):
    """
    This function takes in 2 vectors of ints or floats representing x and y respectively.
    And returns a list containing the gradient (m) in the first index and y-intercept (b) in the second index.
        
    The formula used is from Khan Academy - https://www.youtube.com/watch?v=GAmzwIkGFgE
    """
    # Size of dataset
    N_x = len(xs)
    N_y = len(ys)

    # Calculate the average of x and y
    avg_x = sum(xs)/N_x
    avg_y = sum(ys)/N_y

    # Calculate the average of a list of xys
    xys = [xs[i]*ys[i] for i in range(N_x)]
    avg_xy = sum(xys)/N_x

    # The square of average of x
    squared_avg_x = pow(avg_x, 2)

    # Calculate the average of a list of xs squared
    xs_squared = [pow(x, 2) for x in xs]
    avg_xs_squared = sum(xs_squared)/N_x

    # Calculate the gradient
    m = ((avg_x * avg_y) - avg_xy )/( squared_avg_x - avg_xs_squared )
    b = (avg_y - (m * avg_x))

    return (m, b)


if __name__ == "__main__":
    x = [1,2,3,4,5]
    y = [1,2,3,4,5]

    m, b = linearRegression(x, y)

    print(f"y={m}x+{b}")
